contracts/RelayerHub.sol

```diff
@@ -8,88 +8,150 @@ import "./System.sol";
-    relayer memory r = relayers[msg.sender];
-    msg.sender.transfer(r.deposit.sub(r.dues));
-    address payable systemPayable = address(uint160(SYSTEM_REWARD_ADDR));
-    systemPayable.transfer(r.dues);
-    delete relayersExistMap[msg.sender];
-    delete relayers[msg.sender];
-    emit relayerUnRegister(msg.sender);
-  }
-
-  /*********************** Param update ********************************/
-  function updateParam(string calldata key, bytes calldata value) external override onlyInit onlyGov{
-    if (Memory.compareStrings(key,"requiredDeposit")) {
-      require(value.length == 32, "length of requiredDeposit mismatch");
-      uint256 newRequiredDeposit = BytesToTypes.bytesToUint256(32, value);
-      require(newRequiredDeposit > 1 && newRequiredDeposit <= 1e21 && newRequiredDeposit > dues, "the requiredDeposit out of range");
-      requiredDeposit = newRequiredDeposit;
-    } else if (Memory.compareStrings(key,"dues")) {
-      require(value.length == 32, "length of dues mismatch");
-      uint256 newDues = BytesToTypes.bytesToUint256(32, value);
-      require(newDues > 0 && newDues < requiredDeposit, "the dues out of range");
-      dues = newDues;
-    } else {
-      require(false, "unknown param");
+contract RelayerHub is IRelayerHub, System, IParamSubscriber {
+    using SafeMath for uint256;
+
+    uint256 public constant INIT_REQUIRED_DEPOSIT = 1e20;
+    uint256 public constant INIT_DUES = 1e17;
+    address public constant WHITELIST_1 = 0xb005741528b86F5952469d80A8614591E3c5B632;
+    address public constant WHITELIST_2 = 0x446AA6E0DC65690403dF3F127750da1322941F3e;
+
+    uint256 public requiredDeposit; // have to keep it to not break the storage layout
+    uint256 public dues;
+
+    mapping(address => relayer) relayers; // old map holding the relayers which are to be allowed safe exit
+    mapping(address => bool) relayersExistMap;
+
+    struct relayer {
+        uint256 deposit;
+        uint256 dues;
     }
-    emit paramChange(key, value);
-  }
 
-  function isRelayer(address sender) external override view returns (bool) {
-    return relayersExistMap[sender];
-  }
+    mapping(address => bool) relayManagersExistMap;
+    mapping(address => address) managerToRelayer;
+    mapping(address => bool) currentRelayers;
+
+    bool public whitelistInitDone;
+
+    modifier onlyManager(){
+        require(relayManagersExistMap[msg.sender], "manager does not exist");
+        _;
+    }
+
+    modifier exist() {
+        require(relayersExistMap[msg.sender], "relayer do not exist");
+        _;
+    }
+
+    event relayerRegister(address _relayer);
+    event relayerUnRegister(address _relayer);
+    event paramChange(string key, bytes value);
+
+    event removeManagerEvent(address _removedManager);
+    event addManagerByGovEvent(address _addedManager);
+    event registerManagerEvent(address _registeredManager);
+    event updateRelayerEvent(address _from, address _to);
+
+    function init() external onlyNotInit {
```

Comment:
We should not modify the init function. As the init function will never be called again.